master/addallocationstrigger.sql

```diff
@@ -0,0 +1,171 @@
+-- add sequence numbers to allocations
+ALTER TABLE allocations ADD COLUMN IF NOT EXISTS seq bigint;
+CREATE SEQUENCE IF NOT EXISTS stream_allocation_seq START 1;
+
+-- trigger function to update the sequence number on allocation row modfication
+-- this should fire before so that it can modify new directly
+CREATE OR REPLACE FUNCTION stream_allocation_seq_modify() RETURNS TRIGGER AS $$ 
+BEGIN
+    NEW.seq = nextval('stream_allocation_seq');
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS stream_allocation_trigger_seq ON allocations;
+CREATE TRIGGER stream_allocation_trigger_seq 
+BEFORE INSERT OR UPDATE
+ON allocations
+FOR EACH ROW EXECUTE PROCEDURE stream_allocation_seq_modify();
+
+-- helper function to create allocation jsonb object for streaming 
+CREATE OR REPLACE FUNCTION stream_allocation_notify(
+    before jsonb, beforework integer, after jsonb, afterwork integer, eid integer
+) RETURNS integer AS $$
+DECLARE
+    output jsonb = NULL;
+    temp jsonb = NULL;
+BEGIN
+    IF before IS NOT NULL THEN
+        temp = before || jsonb_object_agg('workspace_id', beforework);
+        temp = before || jsonb_object_agg('experiment_id', eid);
+        output = jsonb_object_agg('before', temp);
+    END IF;
+    IF after IS NOT NULL THEN 
+        temp = after || jsonb_object_agg('workspace_id', afterwork);
+        temp = after || jsonb_object_agg('experiment_id', eid);
+        IF output IS NULL THEN
+            output = jsonb_object_agg('after', temp);
+        ELSE 
+            output = output || jsonb_object_agg('after', temp);
+        END IF;
+    END IF;
+    PERFORM pg_notify('stream_allocation_chan', output::text);
+    return 0;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger function to NOTIFY the master of changes, for INSERT
+CREATE OR REPLACE FUNCTION stream_allocation_change() RETURNS TRIGGER AS $$ 
+DECLARE
+    tid  integer;
+    eid  integer;
+    proj integer;
+    work integer;
+BEGIN
+    IF (TG_OP = 'INSERT') THEN 
+        tid = trial_id from trial_id_task_id WHERE trial_id_task_id.task_id = NEW.task_id;
+        eid = experiment_id FROM trials WHERE trials.id = tid;
+        proj = project_id from experiments where experiments.id = eid;
+        work = workspace_id from projects where projects.id = proj;
+        PERFORM stream_allocation_notify(NULL, NULL, to_jsonb(NEW), work, eid);
+    ELSEIF (TG_OP = 'UPDATE') THEN
+        tid = trial_id from trial_id_task_id WHERE trial_id_task_id.task_id = NEW.task_id;
+        eid = experiment_id FROM trials WHERE trials.id = tid;
+        proj = project_id from experiments where experiments.id = eid;
+        work = workspace_id from projects where projects.id = proj;
+        PERFORM stream_allocation_notify(to_jsonb(OLD), work, to_jsonb(NEW), work, eid);
+    ELSEIF (TG_OP = 'DELETE') THEN
+        tid = trial_id from trial_id_task_id WHERE trial_id_task_id.task_id = OLD.task_id;
+        eid = experiment_id FROM trials WHERE trials.id = tid;
```

Comment:
this also gets messy because experiments have a `DELETING` status they get put into prior to actually being deleted

Reply:
this could be a case where we call bootem, but I don't like the idea of having an `experiment` being deleted causing a permission scope change

Reply:
Should an experiment deletion look like a deletion for an allocation? Since that's not the behavior the DB is exhibiting and we're keen on streaming rows from the DB, I want to default to the DB behavior (especially since the information would still be relevant even if the experiment is no longer there)

Reply:
The permission logic for allocations needs work because it's not workspace-based, it's experiment-based; so there might not be need to get the workspace here at all...but that's messy at best; it looks likes there isn't a permission check on getting an allocation; just editing them...

Reply:
I just wrote the view allocation permission logic (have a PR up for [here](https://github.com/determined-ai/determined/pull/8281))
 
The permission check is workspace-based, but based on two different permission depending on in the allocation is from an NTSC or an Experiment. So will probably have two different access scopes, one for each allocation source. 

Only challenge is efficiently getting the workspace into the allocation event on the SQL side; currently it goes through the experiment, which will not always be possible.